internal/hammer/loadtest/workers.go

```diff
@@ -231,3 +273,178 @@ func (w *LogWriter) Kill() {
 		w.cancel()
 	}
 }
+
+// NewMMDVerifier creates a MMDVerifier.
+func NewMMDVerifier(tracker *client.LogStateTracker, errChan chan<- error, leafMMDChan chan LeafMMD, mmdDuration time.Duration) *MMDVerifier {
+	return &MMDVerifier{
+		tracker:     tracker,
+		errChan:     errChan,
+		leafMMDChan: leafMMDChan,
+		mmdDuration: mmdDuration,
+	}
+}
+
+// MMDVerifier verifies the signed timestamp against the MMD policy for newly
+// added entries by performing inclusion proof.
+type MMDVerifier struct {
+	tracker     *client.LogStateTracker
+	errChan     chan<- error
+	leafMMDChan chan LeafMMD
+	mmdDuration time.Duration
+	cancel      func()
+}
+
+// Run runs the MMD verifier. This should be called in a goroutine.
+func (v *MMDVerifier) Run(ctx context.Context) {
+	if v.cancel != nil {
+		panic("MMDVerifier was ran multiple times")
+	}
+	ctx, v.cancel = context.WithCancel(ctx)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+		}
+
+		var leafMMD *LeafMMD
+		for {
+			if leafMMD == nil {
+				mmd, ok := <-v.leafMMDChan
+				if !ok {
+					break
+				}
+				leafMMD = &mmd
+			}
+
+			// Retry if the leaf is not yet integrated into the log.
+			if leafMMD.index >= v.tracker.LatestConsistent.Size {
+				// Verify MMD timestamp.
+				sec := int64(leafMMD.timestamp / 1000)
+				nsec := int64((leafMMD.timestamp % 1000) * 1000000)
+				if time.Unix(sec, nsec).Add(v.mmdDuration).Before(time.Now()) {
+					v.errChan <- fmt.Errorf("leaf index %d MMD violation at %d", leafMMD.index, leafMMD.timestamp)
+				}
+
+				break
+			}
+
+			// A new proof builder is required because v.tracker.ProofBuilder
+			// is not thread safe.
+			size := v.tracker.LatestConsistent.Size
+			hash := v.tracker.LatestConsistent.Hash
+			pb, err := client.NewProofBuilder(ctx, log.Checkpoint{
+				Origin: v.tracker.Origin,
+				Size:   size,
+				Hash:   hash,
+			}, v.tracker.TileFetcher)
+			if err != nil {
+				v.errChan <- fmt.Errorf("failed to create proof builder: %w", err)
```

Comment:
Again with all of these errors, you need to consider whether `leafMMD` should be nilled out, or whether you think it might pass on subsequent attempts. As written, this looks like it could get stuck in a death loop given a bad entry. Maybe that's a feature, maybe that's a bug.

Reply:
Updated. I'm not sure if we shoud `panic` or fail with fatal log. It is expected to always build the `ProofBuilder` successfully.